MapleServer2/Types/MushkingRoyaleQueue.cs

```diff
@@ -0,0 +1,68 @@
+﻿using MapleServer2.Managers;
+using MapleServer2.Packets;
+using MapleServer2.Servers.Game;
+
+namespace MapleServer2.Types;
+
+public class MushkingRoyaleQueue
+{
+    public List<Player> Players;
+    private Task MatchMaking;
+    private int LastJoinTick;
+
+    public MushkingRoyaleQueue()
+    {
+        Players = new();
+    }
+
+    public void AddPlayer(Player player)
+    {
+        Players.Add(player);
+        MatchMaking ??= MatchMakeLoop();
+        LastJoinTick = Environment.TickCount;
+    }
+
+    public void RemovePlayer(Player player)
+    {
+        Players.Remove(player);
+        if (Players.Count <= 0)
+        {
+            MatchMaking = null;
+        }
+    }
+
+    private Task MatchMakeLoop()
+    {
+        return Task.Run(async () =>
+        {
+            while (Players.Count > 0)
+            {
+                // a match needs minimum of 20 to start. doing 25 to be safe incase of drop outs
+                // TODO: Handle getting more players in while it's still open
+                // TODO: Maybe royale level matching?
+                if (Environment.TickCount > LastJoinTick + 60000 && Players.Count >= 25)
+                {
+                    StartMatch();
+                }
+
+                if (Players.Count >= 50)
+                {
+                    StartMatch();
+                }
+                await Task.Delay(10000); // 10 second delay
+            }
+        });
+    }
+
+    private void StartMatch()
+    {
+        List<Player> players = Players.Take(50).ToList();
```

Comment:
I think using an actual queue here is best , at least you will get the players in order of join which is nice

Reply:
Doesn't Take do this?
https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.take?view=net-6.0

newer players in the queue would be much lower in the queue

Reply:
Yeah I guess it didn't matter